clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -561,6 +584,91 @@ scheduler_tmpl_t<RecordType, ReaderType>::init(
     return STATUS_SUCCESS;
 }
 
+template <typename RecordType, typename ReaderType>
+std::string
+scheduler_tmpl_t<RecordType, ReaderType>::recorded_schedule_component_name(
+    output_ordinal_t output)
+{
+    static const char *const SCHED_CHUNK_PREFIX = "output.";
+    std::ostringstream name;
+    name << SCHED_CHUNK_PREFIX << std::setfill('0') << std::setw(4) << output;
+    return name.str();
+}
+
+template <typename RecordType, typename ReaderType>
+typename scheduler_tmpl_t<RecordType, ReaderType>::scheduler_status_t
+scheduler_tmpl_t<RecordType, ReaderType>::write_recorded_schedule()
+{
+    if (options_.record_ostream == nullptr)
+        return STATUS_ERROR_INVALID_PARAMETER;
+    std::lock_guard<std::mutex> guard(sched_lock_);
+    for (int i = 0; i < static_cast<int>(outputs_.size()); i++) {
+        sched_type_t::stream_status_t status =
+            record_schedule_segment(i, schedule_record_t::FOOTER, 0, 0, 0);
+        if (status != sched_type_t::STATUS_OK)
+            return STATUS_ERROR_FILE_WRITE_FAILED;
+        std::string name = recorded_schedule_component_name(i);
+        std::string err = options_.record_ostream->open_new_component(name);
+        if (!err.empty()) {
+            VPRINT(this, 1, "Failed to open component %s in record file: %s\n",
+                   name.c_str(), err.c_str());
+            return STATUS_ERROR_FILE_WRITE_FAILED;
+        }
+        if (!options_.record_ostream->write(
+                reinterpret_cast<char *>(outputs_[i].record.data()),
+                outputs_[i].record.size() * sizeof(outputs_[i].record[0])))
+            return STATUS_ERROR_FILE_WRITE_FAILED;
+    }
+    return STATUS_SUCCESS;
+}
+
+template <typename RecordType, typename ReaderType>
+typename scheduler_tmpl_t<RecordType, ReaderType>::scheduler_status_t
+scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
+{
+    if (options_.replay_istream == nullptr)
+        return STATUS_ERROR_INVALID_PARAMETER;
+
+    schedule_record_t record;
+    // We assume we can easily fit the whole context switch sequence in memory.
+    // If that turns out not to be the case for very long traces, we deliberately
+    // used an archive format so we could do parallel incremental reads.
+    // (Conversely, if we want to commit to storing in memory, we could use a
+    // non-archive format and store the output ordinal in the version record.)
+    for (int i = 0; i < static_cast<int>(outputs_.size()); i++) {
+        std::string err =
+            options_.replay_istream->open_component(recorded_schedule_component_name(i));
+        if (!err.empty()) {
+            error_string_ = "Failed to open replay_istream component " +
+                recorded_schedule_component_name(i) + ": " + err;
+            return STATUS_ERROR_INVALID_PARAMETER;
+        }
+        // XXX: This could be made more efficient if we stored the record count
```

Comment:
Well, std::vector does some smart sizing internally right?

Reply:
It presumably doubles each time so it won't be too bad, right, and this is just a one-time thing.